### ISSUE TITLE
Merge Devanagari script

### ISSUE THREAD

**Issue by chrissimpkins** [googlefonts/googlesans#148]
Tracking thread for the Devanagari script merge

**Comment by kalapi** [comment]
Thanks @chrissimpkins

Hello @madig and @belluzj! Nice to see you here, I hope you are well and safe! :)

I'm writing to enquire if it's possible to get more time to send you the final sources for testing? I'm aware that you have a Devanagari specialist scheduled to review the work soon which is why I wanted to know how flexible you are with regards to receiving the files. We estimate about 2-3 days worth of work still remaining and I want to be certain that we send you fully functioning sources.

Please let me know what you think, best– Kalapi

**Comment by belluzj** [comment]
Hello Kalapi, nice to see you too! Same good wishes to you.

On our side we have someone booked to merge Indic feature code from your delivery into the main Google Sans project from Wednesday to Friday inclusive (it's Jonny Pinhorn by the way). Before Jonny can work on the features, Nikolaus will need to transfer the glyphs and the kerning using his automated script, which could be done tomorrow afternoon at the latest.

It would be great to know what sort of work remains to be done on your side. For example, I imagine that if you still have to write large portions of feature code, then Jonny's booking should be postponed, or we should find someone else internally to do the work next week onwards, because there would be no point merging a half-working font this week.

However if by tomorrow mid-day you can have final glyph and kerning data, and by Wednesday morning final feature code, we can work with that. If you need to perform final testing on your end on Wednesday onwards, that's fine. It would still be manageable for us if your team has to send a few small bugfixes over the end of the week such as "add these few lines in this lookup" etc.

How does that sound?

**Comment by kalapi** [comment]
Hey!

Sounds good, i'll have the files ready for you by mid-day tomorrow. 

The feature code is all complete, it's small edge-case niggles which are taking time, but I suppose these could be incorporated in a future update. I'll prioritise to make sure that the shaping covers 95% use-case. 

Thanks for a prompt response! Best–kalapi

**Comment by kalapi** [comment]
Hello @madig & @belluzj,

I'm attaching the Devanagari source file containing only the uprights (min & max opsz). I need another hour or two to send you the Italics, I wonder if in the meantime, you can start testing this sources.

Based on your offer I will send you the final sources with the latest shaping information before the start of your day tomorrow.

Please let me know if you have any questions. 

[GoogleSans_Devanagari.zip](https://github.com/googlefonts/googlesans/files/5951299/GoogleSans_Devanagari.zip)

P.S.: Apologies in advance for not following the vendor guidelines and sending you the entire zipped repo, i'll do this tomorrow when I send you the entire package.

**Comment by madig** [comment]
Hi Kalapi! Can you please provide me with a list of glyphs and kerning groups to import? Can't really proceed without one.

**Comment by kalapi** [comment]
~~Hi Nikolaus, I'm confused. What format do you want these in? Is there any documentation I can refer to?~~

Found it, thanks!

**Comment by madig** [comment]
Try this: https://github.com/googlefonts/googlesans/blob/eddd79717294fb2a4f6b6d837502eda10d397a7c/docs/MAINTAINER.md#getting-a-list-of-glyphs-and-kerning-groups

**Comment by kalapi** [comment]
The lists you requested are under the `docs` directory. 

[GoogleSans_Devanagari.zip](https://github.com/googlefonts/googlesans/files/5952822/GoogleSans_Devanagari.zip)

**Comment by madig** [comment]
> ERROR: Glyph ''braceright.loclDEVA'' does not exist in the source UFO GoogleSansDevanagari-Regular.ufo, aborting.

**Comment by kalapi** [comment]
Did you get this when running `gs-glyphs2ufo.py`? Because it parses for me.

**Comment by madig** [comment]
My file was outdated, werks now.

**Comment by madig** [comment]
Oh, you're using the `rclt` feature :D Does that werk in InDesign? @chrissimpkins

**Comment by madig** [comment]
> fontmake: Error: In 'GoogleSans/GoogleSans.designspace': Generating fonts from Designspace failed: /home/nikolaus/googlesans/source/GoogleSans/devanagari.fea:34:36: The following glyph names are referenced but are missing from the glyph set: k_ra-deva

Same for g_ra-deva.

**Comment by kalapi** [comment]
I tested the merge process here (until before combining the features) and it was successful so far. I'm going to continue on the Italic now and will continue testing the merge later. 

> Oh, you're using the `rclt` feature :D Does that werk in InDesign? @chrissimpkins

I'm using `rclt` to capture a cross-cluster context. In the bottom example, it's the _anusvara_ (dot accent) on the previous cluster which I use as context to swap a narrower _eMatra_.

![Screenshot 2021-02-09 at 11 53 26 PM](https://user-images.githubusercontent.com/15804460/107409329-0d438e00-6b32-11eb-8595-98bdcbf43a00.png)

**Comment by kalapi** [comment]
> > fontmake: Error: In 'GoogleSans/GoogleSans.designspace': Generating fonts from Designspace failed: /home/nikolaus/googlesans/source/GoogleSans/devanagari.fea:34:36: The following glyph names are referenced but are missing from the glyph set: k_ra-deva
> 
> Same for g_ra-deva.

Ah, this is part of the 'disabled' code. I'll clean this.

**Comment by belluzj** [comment]
@chrissimpkins Could you please add @jonpinhorn to the repository? He will be working on merging the feature files.

**Comment by davelab6** [comment]
Jonny has been invited.

**Comment by madig** [comment]
Hi Kalapi, do you have any functional feature code for me? We can't proceed unless we can fully build _some_ fonts.

**Comment by kalapi** [comment]
Hi @madig and @jonpinhorn,

Apologies for the delay. Here is the latest package, txt files are in the `docs` directory along with a web-tester we use to test shaping. 

[GoogleSans_Devanagari.zip](https://github.com/googlefonts/googlesans/files/5957621/GoogleSans_Devanagari.zip)

Please let me know if you have any questions.

**Comment by madig** [comment]
Thanks! I noticed that the feature code includes explicit positioning in `dev2_dflt_mar_nep_abvm_contextual_reph_shift_b` and `dev2_dflt_blwm_reposition_nukta_with_rakaar`. This is not going to interpolate except in the VF @chrissimpkins @moyogo Same goes for the adjustments to Hebrew btw.

**Comment by kalapi** [comment]
@madig I'm still testing this. My understanding is that if this is included under the `Replace Feature` custom parameter in the Masters panel, glyphLib interpolates the values. 

We can remove this for now since it affects only one glyph (sorry for not omitting it before sending it across).

**Comment by madig** [comment]
Ok, I'll leave it out of my merge PR for now. It is possible to hack this somehow, but I'd need to think about it.

**Comment by madig** [comment]
> fontmake: Error: In 'GoogleSans/GoogleSans-Italic.designspace': Interpolating instance 'Medium Italic' failed.: Failed to generate instance of glyph 'eyelashReph-deva': list index out of range. (Note: the most common cause for an error here is that the glyph outlines are not point-for-point compatible or have the same starting point or are in the same order in all masters.)

**Comment by kalapi** [comment]
How do you want to go about resolving these issues? Should I fix them and send you updated sources?

**Comment by madig** [comment]
Yas please 🙇‍♀️

**Comment by kalapi** [comment]
Here you go: [GoogleSans_Devanagari.zip](https://github.com/googlefonts/googlesans/files/5958055/GoogleSans_Devanagari.zip)

**Comment by madig** [comment]
If you can, sure!

**Comment by madig** [comment]
@chrissimpkins I just got saved by the regression suite. Kalapi's sources will overwrite various existing kerning groups, wiping out Armenian additions. This happened to change kerning, which was caught by HarfBuzz :)

**Comment by chrissimpkins** [comment]
> @chrissimpkins I just got saved by the regression suite. Kalapi's sources will overwrite various existing kerning groups, wiping out Armenian additions. This happened to change kerning, which was caught by HarfBuzz :)

🥳 🚀 !

**Comment by madig** [comment]
@kalapi can you tell me more about the state of your work? What are you working on still, what is missing? Also, do the fonts in the web test asset folder reflect the current state of your work?

**Comment by kalapi** [comment]
@madig This is pretty much the baseline release. Because this font will be used everywhere, we've been trying to add edge-case shaping support to cover all possible cluster formation. As part of this we will continue to add/modify shaping code as well as add/modify a few existing glyphs. 

Yes, the web-tester reflects the current state of support. 

I'm guessing you're asking because of merge complications?

**Comment by madig** [comment]
Ok, yeah, I'm wondering why some anchoring doesn't work when compiling the fonts with fontmake. I was given these strings to test:

क्ष्र ज्ञ्र क़्र
त्रु त्रू त्रृ त्रे त्रै
द्रु द्रू द्रृ द्रे द्रै

top and bottom thingies seem torn off. See font artifacts at https://github.com/googlefonts/googlesans/pull/186/checks?check_run_id=1879056799

**Comment by kalapi** [comment]
Just checked this, I suspect this is a GDEF issue. Those glyphs are classified as `ligature` when they should be `base`.

The diagonal mark in the third cluster/first line works because the underlying consonant is a `base`.

**Comment by madig** [comment]
I think that did the trick, thanks! I wonder how Glpyhs.app knows these are supposed to be bases...

**Comment by kalapi** [comment]
I think because of these kind of shaping errors Glyphs defaults to classifying these as bases unless explicitly noted otherwise in the GlyphData.xml.

**Comment by madig** [comment]
This might be relevant for glyphLib's GDEF generator @anthrotype. Ideally this is stored directly in the new public.openTypeCategories key.

**Comment by anthrotype** [comment]
i wonder why were they classified as ligatures

**Comment by madig** [comment]
@kalapi Jonny has compiled a Google doc of things he found, what's your email we can share it with?

**Comment by chrissimpkins** [comment]
I'll send an email with all of you on it.

**Comment by chrissimpkins** [comment]
And you can link the Doc in this thread.  Ideally, return to this thread / open new issues for discussion rather than holding this in email.

**Comment by madig** [comment]
See Internal doc

**Comment by jonpinhorn** [comment]
Hello, All.
I'm just going to flag things as I go, in no particular order:

1, र्फ़्रिं Shorter ikaar sub required for all base फ़ + mark combos to be consistent with र्फ्रिं
![image](https://user-images.githubusercontent.com/8477580/107671874-3b3ce580-6c8c-11eb-9578-2e8fece4ff6d.png)

2, /h_m_ya-deva would be required if supporting this conjunct cluster.
![image](https://user-images.githubusercontent.com/8477580/107672272-9e2e7c80-6c8c-11eb-8cf6-57744606dcd9.png)

3, Half-form clashes that could be remedied with positive kerning, though that itself would break the texture. Likely not important.
![image](https://user-images.githubusercontent.com/8477580/107672816-40e6fb00-6c8d-11eb-9699-bffeeac9147e.png)

4, This is more of a question for KG - I was surprised to see श्व not supported in the traditional form - Without it, I'd have thought you'd want this ikaar to stretch and not be the default? Same with र्च्छ्वि that also doesn't have a छ्व glyph. Is there a strategy of unsupported glyphs not having alternate ikaar lengths?
![image](https://user-images.githubusercontent.com/8477580/107673743-2e20f600-6c8e-11eb-9cc5-0440240be7e1.png)

Other stuff that had been in my list has been fixed by Niko correcting the gdef base issue. Cheers.

**Comment by kalapi** [comment]
Hello @jonpinhorn

Many thanks for your review. Here are my thoughts:

1. Will fix this, I haven't added `fa-deva` to the glyph class. 
2. We're not supporting `h_m_ya-deva` (or `c_ch_va` and the traditional `sha` conjuncts) in this character-set since these are Sanskrit shaping conventions and not common in Hindi or Marathi (and other Devanagari script-based) languages. They are in the tester because it was originally developed for Noto Serif Devanagari, so i'll remove those tests for now. If there's a requirement to add the traditional forms in the future we'll do it.
3. Yes this needs to be fixed. Because the Sindhi implosives (`gga-deva`, `jja-deva`, `ddda-deva`, and `bba-deva`) do not usually conjugate with other, non-implosive consonants (other than Ra), and hence these combinations would never occur, but we're taking an 'everything-to-everything' approach this needs to look correct.
4. I'll fix this, don't know why this was skipped. We're using linear combinations for the Sha other than the `sh_ra-deva`.

I'll fix the things that need fixing and resend an update in a few hours. Thanks again!

**Comment by jonpinhorn** [comment]
Hi @kalapi,

Makes perfect sense. Thanks.

![image](https://user-images.githubusercontent.com/8477580/107755299-843b7b00-6d1a-11eb-8e82-0c616ff16837.png)
![image](https://user-images.githubusercontent.com/8477580/107755364-9a493b80-6d1a-11eb-85bb-f82c2bf3953c.png)


I've found that /ueMatra-deva and /uueMatra-deva behave differently between your source and the new one. My understanding is there's no use case except in the composite /ue-deva and /uue-deva - this is fine. But, for some reason your sources still position the un-anchored marks in what looks like a 'bottom' position. Any thoughts on why there could be a difference?

**Comment by kalapi** [comment]
@jonpinhorn This is strange, haha! There are no corresponding anchors in the sources (`ueMatra-deva` uses anchor `_bottomright`). I'll 'fix' this by removing the anchors and manually positioning the matras under the full vowels.

**Comment by kalapi** [comment]
Hello @jonpinhorn & @madig 

I've attached the updated package here (updated the `import_groups<_italic>.txt` too).

Jonny, I've addressed 1 & 4 in this release and I'm leaving 3 for the next release because we have to review some other edge-case kerning too. 

With regards to the second example in 4, i've not incorporated the iMatra contexts for combinations with an explicit rakar. We'll have to look at all the possible combinations with this in the next release.

For the `ueMatra-deva` and `uueMatra-deva`, i've removed the anchor but please not that the shaping will now look like the second example in your reference image. This is because I don't want the below matra clashing with the following glyphs/clusters. It's better for them to roughly position under the glyph. 

Please let me know if you have any questions. Thanks–K

[GoogleSans_Devanagari_R02.zip](https://github.com/googlefonts/googlesans/files/5971559/GoogleSans_Devanagari_R02.zip)

**Comment by madig** [comment]
Thanks! Passing the built fonts on to Jonny.

PS: Can you please remove the couple of glyphs from the import list that are in base already, e.g. CR, space, nbspace, etc.? The list should only contain glpyhs you really want imported.

**Comment by madig** [comment]
@kalapi btw, did you decide on what to do about the abvm and blwm features you tried to use to position some stuff?

**Comment by kalapi** [comment]
@madig I haven't added all of the feature code yet, I plan to do this next week. Do you require test sources to check how the pipeline handles them? I'm afraid we'll have to figure this out at some point because for the Devanagari it is used to support edge-case niceties but in scripts like Telugu this will be the backbone of the shaping.

**Comment by madig** [comment]
Nope, we still have to figure that out. Cosimo had some ideas.

**Comment by moyogo** [comment]
Regarding the position adjustments in the abvm and blwm features, this requires the automatic code insertion marker to work in ufo2ft. Otherwise the automatic code is appended and overrides any position adjustments.

**Comment by madig** [comment]
@kalapi Jonny is done checking the files. Can you please do a round of checks on our merged files as well? https://github.com/googlefonts/googlesans/actions/runs/561809440

**Comment by chrissimpkins** [comment]
Pinging current status as I understand it here.  

Kalapi in https://github.com/googlefonts/googlesans/issues/148#issuecomment-778264769

>  I haven't added all of the feature code yet, I plan to do this next week. Do you require test sources to check how the pipeline handles them? I'm afraid we'll have to figure this out at some point because for the Devanagari it is used to support edge-case niceties but in scripts like Telugu this will be the backbone of the shaping.

Niko in https://github.com/googlefonts/googlesans/issues/148#issuecomment-778267081

> Nope, we still have to figure that out. Cosimo had some ideas.

Denis in https://github.com/googlefonts/googlesans/issues/148#issuecomment-778276456

> Regarding the position adjustments in the abvm and blwm features, this requires the automatic code insertion marker to work in ufo2ft. Otherwise the automatic code is appended and overrides any position adjustments.

**Comment by madig** [comment]
Haven't had time to look into that yet.

**Comment by chrissimpkins** [comment]
> Haven't had time to look into that yet.

@madig Is there anything outstanding in this thread from your standpoint?  I am prepared to merge #277

**Comment by chrissimpkins** [comment]
#304 opened with a handful of issues that we need to address in a future Devanagari script update.  These were based on the visual review of #277 changes.

**Comment by madig** [comment]
ufo2ft 2.20.0 added support for feature insertion markers, so the abvm and blwm features Kalapi envisioned should work for Telugu . I have however no experience with those features, maybe @moyogo knows more? We'll want to update dependencies soon anyway.

I think the Deva PR is ready to go otherwise.
